sdk/core/core/README.md

```diff
@@ -0,0 +1,123 @@
+# Azure SDK Core Library
+
+In high-level programming languages, an [inversion of control](https://en.wikipedia.org/wiki/Inversion_of_control)
+is commonly used. For example, a generic JSON deserializer may call a user provided function to set an object property value
+(similar to [Visitor Pattern](https://en.wikipedia.org/wiki/Visitor_pattern)). The principle requires a call-back
+function or a virtual table. In C, a call-back or a virtual table requires an untyped parameter `void *context`.
+It's very hard to work safely with such functions so Azure SDK Core library for C tries to avoid these technic.
+Instead, the Core library provides safe and useful parts to construct complex solutions.
+
+## 1. Data Types
+
+### 1.1. Const in `struct`
+
+Don't use `const` fields in structures and unions. For example:
+
+```c
+struct my_struct {
+  size_t const size;
+};
+
+// ok
+struct my_struct my_struct_create() {
+  return (struct my_struct){ .size = 5 };
+}
+
+void my_struct_init(struct my_struct *const p) {
+  // compilation error.
+  *p = (struct my_struct){ .size = 5 };
+}
+```
+
+### 1.2. Tagged Unions
+
+Example of `az_mytype` tagged union.
+
+```c
+enum {
+  AZ_MYTYPE_FOO,
+  AZ_MYTYPE_BAR,
+};
+
+typedef uint8_t az_mytype_kind;
+
+typedef double az_mytype_foo;
+
+typedef struct {
+  double a;
+  double b;
+} az_mytype_bar;
+
+inline az_mytype_bar az_mytype_bar_create(double const a, double const b) {
+  return (az_mytype_bar){ .a = a, .b = b };
+}
+
+struct my_type {
```

Comment:
what's this doing here?

Reply:
Yes, we need to move to Design Guidance.